Update `trainer._get_eval_sampler()` to support `group_by_length` arg

larin92 opened the pull request:
Trainer didn't support grouping batches by length for evaluation, which made evaluation slow with `eval_batch_size`>1.

Updated `trainer._get_eval_sampler()` method was based off of `trainer._get_train_sampler()`.

larin92 commented:
checked failed test, it's not related to this PR

SunMarc commented on src/transformers/trainer.py:
```diff
@@ -931,26 +931,27 @@ def get_train_dataloader(self) -> DataLoader:
-            elif is_sagemaker_mp_enabled():
-                return SequentialDistributedSampler(
-                    eval_dataset,
-                    num_replicas=smp.dp_size(),
-                    rank=smp.dp_rank(),
-                    batch_size=self.args.per_device_eval_batch_size,
+        if self.eval_dataset is None or not has_length(self.eval_dataset):
+            return None
+        # Build the sampler.
+        if self.args.group_by_length:
+            if is_datasets_available() and isinstance(self.eval_dataset, datasets.Dataset):
+                lengths = (
+                    self.eval_dataset[self.args.length_column_name]
+                    if self.args.length_column_name in self.eval_dataset.column_names
+                    else None
                 )
             else:
-                return SequentialSampler(eval_dataset)
-
```
We can't just remove this part

larin92 replied:
returned it

SunMarc commented on src/transformers/trainer.py:
```diff
@@ -931,26 +931,27 @@ def get_train_dataloader(self) -> DataLoader:
+                self.args.eval_batch_size,
+                dataset=self.eval_dataset,
+                lengths=lengths,
+                model_input_name=model_input_name,
+            )
         else:
-            return None
```
We should keep that too

larin92 replied:
returned it

SunMarc commented on src/transformers/trainer.py:
```diff
@@ -931,26 +931,27 @@ def get_train_dataloader(self) -> DataLoader:
-            return SequentialSampler(eval_dataset)
+                lengths = None
+            model_input_name = self.tokenizer.model_input_names[0] if self.tokenizer is not None else None
+            return LengthGroupedSampler(
+                self.args.eval_batch_size,
+                dataset=self.eval_dataset,
+                lengths=lengths,
+                model_input_name=model_input_name,
+            )
         else:
```
Did you test this sampler in the eval stage ? How fast it is compared to what we had before ?

larin92 replied:
for my dataset, eval stage:
- `eval_batch_size=8`  current transformers version ~4min, this patched version ~2min

current transformers version of `_get_eval_sampler()` doesn't account for `group_by_length` arg for batching, while `_get_train_sampler()` does account for it

SunMarc reviewed (commented):
Thanks @larin92 ! Left a few comments cc @muellerzr

SunMarc reviewed (approved):
LGTM ! I'm just not sure if we should only use `group_by_length` for both training and evals. cc @muellerzr 

Could you also rebase the PR to main in order to pass the CI ? cc @larin92

muellerzr reviewed (approved):
Thanks! Should be a straightforward speedup for users doing this :) Nice!

larin92 committed:
Update 'trainer._get_eval_sampler()' to support 'group_by_length' argument

Trainer didn't support grouping by length for evaluation, which made evaluation slow with 'eval_batch_size'>1.

Updated 'trainer._get_eval_sampler()' method was based off of 'trainer._get_train_sampler()'.

larin92 commented:
@SunMarc rebased on top of 0f49deacbff3e57cde45222842c0db6375e4fa43 (as latest commit with all checks passed)
